src/rust/ensogl/lib/components/src/selector/common.rs

```diff
@@ -0,0 +1,126 @@
+//! Common functionality for both the Number and Range selector.
+use crate::prelude::*;
+
+use enso_frp as frp;
+use enso_frp::Network;
+use ensogl_core::frp::io::Mouse;
+use ensogl_core::gui::component::ShapeViewEvents;
+
+pub mod base_frp;
+pub mod model;
+pub mod shape;
+
+pub use base_frp::*;
+pub use model::*;
+
+
+
+// ==============
+// === Bounds ===
+// ==============
+
+/// Bounds of a selection. This indicates the lowest and highest value that can be selected in a
+/// selection component.
+pub type Bounds = (f32, f32);
+
+/// Frp utility method to normalise the given value to the given Bounds.
+pub fn normalise_value((value,range):&(f32,Bounds)) -> f32 {
+    (value - range.0) / (range.1 - range.0)
+}
+
+/// Frp utility method to compute the absolute value from a normalised value.
+/// Inverse of `normalise_value`.
+pub fn absolute_value((range,normalised_value):&(Bounds,f32)) -> f32 {
+    ((normalised_value * (range.1 - range.0)) + range.0)
+}
+
+/// Returns the normalised value that correspond to the click posiiton on the shape.
+/// For use in FRP `map` method, thus taking references.
+#[allow(clippy::trivially_copy_pass_by_ref)]
+pub fn position_to_normalised_value(pos:&Vector2,width:&f32) -> f32 {
+    ((pos.x / (width/2.0)) + 1.0) / 2.0
+}
+
+/// Check whether the given value is within the given bounds. End points are exclusive.
+fn value_in_bounds(value:f32, bounds:Bounds) -> bool {
+    value > bounds.0 && value < bounds.1
+}
```

Comment:
I've added tests. Which is good since there was a bug here. Bounds here are meant to be inclusive. So, a range of `[0,0] `is perfectly fine. I've also made sure the functions work with zero-width bounds in a sensible way. Even though NaNs here did not make any problems before.